ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,20 +187,416 @@ let order_f (_, vm_rec) =
+
+let gib_of_bytes x = x |> kib_of_bytes |> kib_of_bytes |> kib_of_bytes
+
+let log_grp_ranked_hosts ~__context grp_ranked_hosts =
+  debug "log_grp_ranked_hosts:" ;
+  VMGroupMap.iter
+    (fun g ranked_hosts ->
+      debug "  group: %s" (Db.VM_group.get_name_label ~__context ~self:g) ;
+      IntMap.iter
+        (fun vm_count hosts ->
+          debug "    vm_count: %d, hosts: [%s]" vm_count
+            (hosts
+            |> List.map (fun h -> Db.Host.get_name_label ~__context ~self:h)
+            |> String.concat ","
+            )
+        )
+        ranked_hosts
+    )
+    grp_ranked_hosts
+
+let log_vms_increasing ~__context vms_increasing =
+  debug "log_vms_increasing:" ;
+  List.iter
+    (fun (vm, size) ->
+      debug "  vm: %s %Ld GB"
+        (Db.VM.get_name_label ~__context ~self:vm)
+        (gib_of_bytes size)
+    )
+    vms_increasing
+
+let log_hosts_increasing ~__context hosts_increasing =
+  debug "log_hosts_increasing:" ;
+  List.iter
+    (fun (h, size) ->
+      debug "  host: %s %Ld GB"
+        (Db.Host.get_name_label ~__context ~self:h)
+        (gib_of_bytes size)
+    )
+    hosts_increasing
+
+let vm_memory ~__context snapshot =
+  let policy =
+    match Helpers.check_domain_type snapshot.API.vM_domain_type with
+    | `hvm | `pv ->
+        Memory_check.Dynamic_min
+    | `pv_in_pvh | `pvh ->
+        Memory_check.Static_max
+  in
+  total_memory_of_vm ~__context policy snapshot
+
+(** Generate 2 layers Map: the outer layer is a Map from `VM_group to ranked_hosts, while each
+    ranked_hosts is a Map from vm_count to `host List, where every host in the list has the
+    same number of running anti-affinity VMs resident on it: vm_count. *)
+let anti_affinity_grp_ranked_hosts ~__context hosts =
+  Db.VM_group.get_all ~__context
+  |> List.filter (fun group ->
+         Db.VM_group.get_placement ~__context ~self:group = `anti_affinity
+     )
+  |> List.fold_left
+       (fun grp_to_ranked_hosts_map group ->
+         VMGroupMap.add group
+           (Xapi_vm_helpers.rank_hosts_by_vm_cnt_in_group ~__context group hosts
+           |> List.fold_left
+                (fun vm_cnt_to_hosts_map hs ->
+                  match hs with
+                  | (_, vm_cnt) :: _ ->
+                      IntMap.add vm_cnt (List.map fst hs) vm_cnt_to_hosts_map
+                  | [] ->
+                      vm_cnt_to_hosts_map
+                )
+                IntMap.empty
+           )
+           grp_to_ranked_hosts_map
+       )
+       VMGroupMap.empty
+
+(** Update the 2 layers Map - grp_ranked_hosts when a new VM from the anti-affinity group is
+    planned to run on the host. *)
+let update_anti_affinity_grp_ranked_hosts group host grp_ranked_hosts =
+  let ranked_hosts =
+    try VMGroupMap.find group grp_ranked_hosts
+    with Not_found ->
+      error "Failed to update grp_ranked_hosts, group not in it" ;
+      raise Not_found
+  in
+  let vm_cnt_on_host =
+    try
+      IntMap.find_first
```

Comment:
`where f is a monotonically increasing function`.

Reply:
I think this is not a problem, for any host, we have one and only one "rank list" which contains it.

Reply:
To make sure it works, changed to use "IntMap.bindings |> List.find" instead, thanks!